editor/src/templates/editor.tsx

```diff
@@ -471,36 +471,118 @@ export class Editor {
-            ],
-            this.storedState,
-            this.spyCollector,
-          )
-          this.storedState = dispatchResultWithMetadata
-          entireUpdateFinished = Promise.all([
-            entireUpdateFinished,
-            dispatchResultWithMetadata.entireUpdateFinished,
-          ])
+          if (domWalkerResult != null) {
+            const dispatchResultWithMetadata = editorDispatch(
+              this.boundDispatch,
+              [
+                EditorActions.saveDOMReport(
+                  domWalkerResult.metadata,
+                  domWalkerResult.cachedPaths,
+                  domWalkerResult.invalidatedPaths,
+                ),
+              ],
+              this.storedState,
+              this.spyCollector,
+            )
+            this.storedState = dispatchResultWithMetadata
+            entireUpdateFinished = Promise.all([
+              entireUpdateFinished,
+              dispatchResultWithMetadata.entireUpdateFinished,
+            ])
+          }
+        }
+
+        // true up groups if needed
+        if (this.storedState.unpatchedEditor.trueUpGroupsForElementAfterDomWalkerRuns.length > 0) {
+          // updated editor with trued up groups
+          Measure.taskTime(`Group true up ${updateId}`, () => {
+            const projectContentsBeforeGroupTrueUp =
+              this.storedState.unpatchedEditor.projectContents
+            const dispatchResultWithTruedUpGroups = editorDispatch(
+              this.boundDispatch,
+              [EditorActions.mergeWithPrevUndo([{ action: 'TRUE_UP_GROUPS' }])],
+              this.storedState,
+              this.spyCollector,
+            )
+            this.storedState = dispatchResultWithTruedUpGroups
+
+            entireUpdateFinished = Promise.all([
+              entireUpdateFinished,
+              dispatchResultWithTruedUpGroups.entireUpdateFinished,
+            ])
+
+            if (
+              projectContentsBeforeGroupTrueUp === this.storedState.unpatchedEditor.projectContents
+            ) {
+              // no group-related re-render / re-measure is needed, bail out
+              return
+            }
+
+            // re-render the canvas
+            Measure.taskTime(`Canvas re-render because of groups ${updateId}`, () => {
+              ElementsToRerenderGLOBAL.current = fixElementsToRerender(
+                this.storedState.patchedEditor.canvas.elementsToRerender,
+                dispatchedActions,
+              ) // Mutation!
+
+              ReactDOM.flushSync(() => {
+                ReactDOM.unstable_batchedUpdates(() => {
+                  this.canvasStore.setState(
+                    patchedStoreFromFullStore(this.storedState, 'canvas-store'),
+                  )
+                })
+              })
+            })
+
+            // re-run the dom-walker
+            Measure.taskTime(`Dom walker re-run because of groups ${updateId}`, () => {
+              const domWalkerResult = runDomWalker({
```

Comment:
I suspect we should extract this block of code since it's near enough identical to the code above, so that if we need to make a change to the process we can ensure it's made in both places

Reply:
good point, done